seed/php-sdk/imdb/src/Core/SerializableType.php

```diff
@@ -0,0 +1,386 @@
+use Seed\Core\SerializableType\UnionArrayType;
+
+/**
+ * Abstract class SerializableType
+ *
+ * Provides generic serialization and deserialization methods.
+ */
+abstract class SerializableType
+{
+    /**
+     * Serializes the object to a JSON string.
+     *
+     * @return string
+     */
+    public function toJson(): string
+    {
+        $arrayData = $this->toArray();
+        return json_encode($arrayData, JSON_THROW_ON_ERROR);
+    }
+
+    /**
+     * Serializes the object to an array for JSON encoding.
+     *
+     * @return array
+     */
+    public function toArray(): array
+    {
+        $result = [];
+        $reflectionClass = new \ReflectionClass($this);
+
+        foreach ($reflectionClass->getProperties() as $property) {
+            $jsonKey = self::getJsonKey($property);
+            $value = $property->getValue($this);
+
+            // Handle DateTime properties
+            $dateTypeAttr = $property->getAttributes(DateType::class)[0] ?? null;
+            if ($dateTypeAttr && $value instanceof DateTime) {
+                $dateType = $dateTypeAttr->newInstance()->type;
+                $value = ($dateType === DateType::TYPE_DATE)
+                    ? $value->format('Y-m-d')
+                    : $value->format(DateTime::RFC3339);
+            }
+
+            // Handle arrays with type annotations
+            $arrayTypeAttr = $property->getAttributes(ArrayType::class)[0] ?? null;
+            if ($arrayTypeAttr && is_array($value)) {
+                $arrayType = $arrayTypeAttr->newInstance()->type;
+                $value = self::serializeGenericArray($value, $arrayType);
+            }
+
+            if ($value !== null) {
+                $result[$jsonKey] = $value;
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * Serializes the given array based on the type annotation.
+     *
+     * @param array $data The array to be serialized.
+     * @param array $type The type definition from the annotation.
+     * @return array The serialized array.
+     */
+    private static function serializeGenericArray(array $data, array $type): array
+    {
+        return self::isMapType($type)
+            ? self::serializeMap($data, $type)
+            : self::serializeList($data, $type);
+    }
+
+    /**
+     * Serializes a value based on the type definition.
+     *
+     * @param mixed $data The value to serialize.
+     * @param mixed $type The type definition.
+     * @return mixed The serialized value.
+     */
+    private static function serializeValue($data, $type)
+    {
+        if ($type instanceof Union) {
+            foreach ($type->types as $unionType) {
+                try {
+                    return self::serializeSingleValue($data, $unionType);
+                } catch (Exception $e) {
+                    continue;
+                }
+            }
+            throw new \InvalidArgumentException("Cannot serialize value with any of the union types.");
+        }
```

Comment:
Does this guarantee that the correct discriminant is set, even if multiple union types have the same underlying form? For example, what if I had a union like:

```yaml
types:
  Name: string
  Union:
     s: string
     name: Name
```

If I specify a `Name`, I want the JSON representation to be:

```json
{
  "type": "name",
  "name": "..."
}
```

But if `Name` is recognizable as a `string`, then would `string` succeed first?

---

Normally we use the discriminant to determine how we should serialize and only attempt that. Maybe that's what is happening here already?

Reply:
Also note if this ends up not being the case, we can always come back to unions later - most important thing to focus on in the first pass is objects and enums.

Reply:
This code will only be reusable for undiscriminated unions, similar to C#

We'll have to figure out how we wanna handle discriminated unions, maybe something similar to Java's